File: BinHeap.java
package src;

import java.util.ArrayList;
import java.util.List;
import java.util.NoSuchElementException;

public class BinHeap {
    private List<Integer> heapvector;  // Список для хранения элементов кучи
    private int currentSize;           // Текущее количество элементов в куче

    // Конструктор класса
    public BinHeap() {
        this.heapvector = new ArrayList<>();
        this.heapvector.add(0); // Фиктивный элемент для индексации с 1
        this.currentSize = 0;
    }

    // Конструктор с начальными значениями
    public BinHeap(List<Integer> initialVector) {
        this();
        this.buildheap(initialVector);
    }

    // Метод просеивания вверх
    private void percUp(int i) {
        while (i > 1) {  // Пока не дошли до корня
            int parentIdx = i / 2;  // Вычисляем индекс родителя
            if (this.heapvector.get(i) < this.heapvector.get(parentIdx)) {
                // Если текущий элемент меньше родителя - меняем местами
                swap(i, parentIdx);
                i = parentIdx;  // Переходим к родительской позиции
            } else {
                break;  // Свойство кучи восстановлено - выходим
            }
        }
    }

    // Вспомогательный метод для обмена элементов
    private void swap(int i, int j) {
        int temp = this.heapvector.get(i);
        this.heapvector.set(i, this.heapvector.get(j));
        this.heapvector.set(j, temp);
    }

    // Добавление элемента в кучу
    public void insert(int k) {
        this.heapvector.add(k);  // Добавляем элемент в конец списка
        this.currentSize++;      // Увеличиваем счетчик элементов
        this.percUp(this.currentSize); // Восстанавливаем свойства кучи
    }

    // Поиск минимального потомка
    private int minChild(int i) {
        int left = i * 2;      // Индекс левого потомка
        int right = i * 2 + 1; // Индекс правого потомка

        // Если правого потомка нет, возвращаем левого
        if (right > this.currentSize) {
            return left;
        }
        // Иначе возвращаем меньшего из двух
        else {
            return (this.heapvector.get(left) < this.heapvector.get(right)) ? left : right;
        }
    }

    // Метод просеивания вниз
    private void percDown(int i) {
        while (i * 2 <= this.currentSize) {  // Пока есть хотя бы левый потомок
            int mc = this.minChild(i);       // Находим минимального потомка

            // Если текущий узел больше наименьшего потомка - меняем
            if (this.heapvector.get(i) > this.heapvector.get(mc)) {
                swap(i, mc);
                i = mc;  // Переходим к позиции потомка
            } else {
                break;  // Свойство кучи восстановлено - выходим
            }
        }
    }

    // Извлечение минимального элемента
    public int delMin() {
        if (this.currentSize == 0) {
            throw new NoSuchElementException("Heap is empty");  // Проверка на пустоту
        }

        int minVal = this.heapvector.get(1);  // Сохраняем минимальный элемент (корень)
        this.heapvector.set(1, this.heapvector.get(this.currentSize));  // Перемещаем последний элемент в корень
        this.heapvector.remove(this.currentSize);  // Удаляем последний элемент
        this.currentSize--;               // Уменьшаем счетчик

        if (this.currentSize > 0) {
            this.percDown(1);  // Восстанавливаем свойства кучи от корня
        }

        return minVal;  // Возвращаем извлеченный минимальный элемент
    }

    // Построение кучи из списка
    public void buildheap(List<Integer> avector) {
        // Очищаем кучу
        this.heapvector = new ArrayList<>();
        this.heapvector.add(0);  // Фиктивный элемент
        this.currentSize = 0;     // Сбрасываем счетчик

        // Добавляем все элементы через insert
        for (int val : avector) {
            this.insert(val);  // Каждый элемент добавляется с восстановлением свойств кучи
        }
    }

    // Построение кучи из массива
    public void buildheap(int[] arr) {
        // Очищаем кучу
        this.heapvector = new ArrayList<>();
        this.heapvector.add(0);  // Фиктивный элемент
        this.currentSize = 0;     // Сбрасываем счетчик

        // Добавляем все элементы через insert
        for (int val : arr) {
            this.insert(val);  // Каждый элемент добавляется с восстановлением свойств кучи
        }
    }

    // Проверка на пустоту
    public boolean isEmpty() {
        return this.currentSize == 0;  // Проверка на пустоту кучи
    }

    // Получение минимального элемента без удаления
    public int findMin() {
        if (this.currentSize == 0) {
            throw new NoSuchElementException("Heap is empty");
        }
        return this.heapvector.get(1);  // Минимальный элемент всегда в корне (индекс 1)
    }

    // Вывод всех элементов кучи
    public void printHeap() {
        System.out.print("Heap elements: ");
        for (int i = 1; i <= this.currentSize; i++) {
            System.out.print(this.heapvector.get(i) + " ");  // Вывод всех элементов кучи
        }
        System.out.println();
    }

    // Получение размера кучи
    public int size() {
        return this.currentSize;
    }

    // Главный метод для тестирования
    public static void main(String[] args) {
        int[] arr = { 3, 1, 23, 5, 6, 7, 43, 8, 864 };  // Исходный массив

        BinHeap bh = new BinHeap();  // Создание объекта кучи
        bh.buildheap(arr);           // Построение кучи из массива

        System.out.print("Initial heap: ");
        bh.printHeap();              // Вывод начальной кучи

        System.out.println("\nExtracting elements:");
        while (!bh.isEmpty()) {      // Пока куча не пуста
            System.out.println("Extracted: " + bh.delMin());  // Извлекаем минимальный элемент
            if (!bh.isEmpty()) {
                System.out.print("Remaining heap: ");
                bh.printHeap();      // Вывод оставшейся кучи
            }
        }              
       
    }
}

File: lab2/BinHeap.cpp
#include <iostream>
#include <stdexcept> // äëÿ ðàáîòû ñ èñêëþ÷åíèÿìè
#include <vector>
using namespace std;

class BinHeap {
private:
    vector<int> heapvector; // Âåêòîð äëÿ õðàíåíèÿ ýëåìåíòîâ êó÷è, heapvector - äèíàìè÷åñêèé ìàññèâ äëÿ õðàíåíèÿ ýëåìåíòîâ
    int CurrentSize; // Òåêóùåå êîëè÷åñòâî ýëåìåíòîâ â êó÷å, CurrentSize - ñ÷åò÷èê ýëåìåíòîâ (èñêëþ÷àÿ ôèêòèâíûé ýëåìåíò)

public:
    BinHeap() {
        this->heapvector.push_back(0); // Ôèêòèâíûé ýëåìåíò äëÿ èíäåêñàöèè ñ 1
        this->CurrentSize = 0;
    }
    // Äîáàâëÿåì ôèêòèâíûé ýëåìåíò 0 â ïîçèöèþ 0
    // Ýòî ïîçâîëÿåò èñïîëüçîâàòü èíäåêñàöèþ ñ 1 äëÿ óäîáñòâà âû÷èñëåíèé
    // Èíèöèàëèçèðóåì ðàçìåð êó÷è êàê 0

    void percUp(int i) {
        while (i > 1) { // Ïîêà íå äîøëè äî êîðíÿ
            int parentIdx = i / 2; // Âû÷èñëÿåì èíäåêñ ðîäèòåëÿ
            if (this->heapvector[i] < this->heapvector[parentIdx]) {
                // Åñëè òåêóùèé ýëåìåíò ìåíüøå ðîäèòåëÿ - ìåíÿåì ìåñòàìè
                swap(this->heapvector[i], this->heapvector[parentIdx]);
                i = parentIdx; // Ïåðåõîäèì ê ðîäèòåëüñêîé ïîçèöèè
            }
            else {
                break; // Ñâîéñòâî êó÷è âîññòàíîâëåíî - âûõîäèì èç öèêëà
            }
        }
    }

    void insert(int k) {
        this->heapvector.push_back(k); // Äîáàâëÿåì ýëåìåíò â êîíåö âåêòîðà è óâåëè÷èâàåì ñ÷åò÷èê
        this->CurrentSize++;
        this->percUp(this->CurrentSize); // Âîññòàíàâëèâàåì ñâîéñòâà êó÷è
    }
    // Âû÷èñëÿåò èíäåêñû ëåâîãî è ïðàâîãî ïîòîìêîâ
    // Âîçâðàùàåò èíäåêñ ïîòîìêà ñ ìåíüøèì çíà÷åíèåì

    int minChild(int i) {
        int left = i * 2; // Èíäåêñ ëåâîãî ïîòîìêà
        int right = i * 2 + 1; // Èíäåêñ ïðàâîãî ïîòîìêà

        // Åñëè ïðàâîãî ïîòîìêà íåò, âîçâðàùàåì ëåâîãî
        if (right > this->CurrentSize) {
            return left;
        }
        // Èíà÷å âîçâðàùàåì ìåíüøåãî èç äâóõ
        else {
            return (this->heapvector[left] < this->heapvector[right]) ? left : right;
        }
    }
    // Âîññòàíàâëèâàåò ñâîéñòâî êó÷è ñâåðõó âíèç
    void percDown(int i) {
        while (i * 2 <= this->CurrentSize) {  // Ïîêà åñòü õîòÿ áû ëåâûé ïîòîìîê
            int mc = this->minChild(i);

            // Åñëè òåêóùèé óçåë áîëüøå íàèìåíüøåãî ïîòîìêà - ìåíÿåì
            if (this->heapvector[i] > this->heapvector[mc]) {
                swap(this->heapvector[i], this->heapvector[mc]);
                i = mc;  // Ïåðåõîäèì ê ïîçèöèè ïîòîìêà
            }
            else {
                break;  // Ñâîéñòâî êó÷è âîññòàíîâëåíî - âûõîäèì
            }
        }
    }

    int delMin() {
        if (this->CurrentSize == 0) { // Ïðîâåðêà íà ïóñòîòó
            throw runtime_error("Heap is empty");
        }

        int minVal = this->heapvector[1]; // Ñîõðàíÿåì ìèíèìàëüíûé ýëåìåíò (êîðåíü)
        this->heapvector[1] = this->heapvector[this->CurrentSize]; // Ïåðåìåùàåì ïîñëåäíèé ýëåìåíò â êîðåíü
        this->heapvector.pop_back(); // Óäàëÿåì ïîñëåäíèé ýëåìåíò
        this->CurrentSize--; // Óìåíüøàåì ñ÷åò÷èê

        if (this->CurrentSize > 0) {
            this->percDown(1);
        }

        return minVal; // Âîçâðàùàåì èçâëå÷åííûé ìèíèìàëüíûé ýëåìåíò
    }

    // Ïîñòðîåíèå êó÷è èç ìàññèâà:
    void buildheap(vector<int> avector) {
        // Î÷èùàåì êó÷ó
        this->heapvector = { 0 }; // Ñîçäàåì íîâûé âåêòîð ñ ôèêòèâíûì ýëåìåíòîì
        this->CurrentSize = 0; // Ñáðàñûâàåì ñ÷åò÷èê

        // Äîáàâëÿåì âñå ýëåìåíòû ÷åðåç insert
        for (int val : avector) {
            this->insert(val); // Êàæäûé ýëåìåíò äîáàâëÿåòñÿ ñ âîññòàíîâëåíèåì ñâîéñòâ êó÷è
        }
    }

    bool isEmpty() {
        return this->CurrentSize == 0;
    }

    void printHeap() { // Âûâîäèò âñå ýëåìåíòû êó÷è (íà÷èíàÿ ñ èíäåêñà 1)
        cout << "Heap elements: ";
        for (int i = 1; i <= this->CurrentSize; i++) {
            cout << this->heapvector[i] << " ";
        }
        cout << endl;
    }
};

int main() {
    int arr[] = { 3, 1, 23, 5, 6, 7, 43, 8, 864 };
    vector<int> a(arr, arr + (sizeof(arr) / sizeof(arr[0])));

    BinHeap* bh = new BinHeap(); // Ñîçäàíèå îáúåêòà êó÷è
    bh->buildheap(a); // Ïîñòðîåíèå êó÷è èç âåêòîðà

    cout << "Initial heap: ";
    bh->printHeap(); // Âûâîä íà÷àëüíîé êó÷è


    cout << "\nExtracting elements:\n";
    while (!bh->isEmpty()) { // Ïîêà êó÷à íå ïóñòà
        cout << "Extracted: " << bh->delMin() << endl; // Èçâëåêàåì ìèíèìàëüíûé ýëåìåíò
        if (!bh->isEmpty()) {
            cout << "Remaining heap: ";
            bh->printHeap(); // Âûâîä îñòàâøåéñÿ êó÷è
        }
    }

    delete bh;
    return 0;
}
